src/Plugin/formatter/FormatterJsonApi.php

```diff
@@ -286,14 +231,80 @@ protected function getRequest() {
   /**
    * Move the embedded resources to the included key.
    *
+   * Change the data structure from an auto-contained hierarchical tree to the
+   * final JSON API structure. The auto-contained tree has redundant information
+   * because every branch contains all the information that is embedded in there
+   * and can be used as stand alone.
+   *
    * @param array $output
    *   The output array to modify to include the compounded documents.
    * @param array $included
    *   Pool of documents to compound.
    *
+   * @return array
+   *   The processed data.
+   *
    * @throws \Drupal\restful\Exception\InternalServerErrorException
    */
-  protected function moveEmbedsToIncluded(array &$output, array $included) {
+  protected function decompress(array $output, array &$included) {
+    static $depth = -1;
+    $depth++;
```

Comment:
Not sure how I feel myself about this `$depth` variable. It's certainly useful for debugging purposes (I use it all the time for conditional breakpoints), but it doesn't drive any business logic.